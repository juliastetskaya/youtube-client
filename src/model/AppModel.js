export default class AppModel {
  constructor(state) {
    this.state = state;
    this.pageToken = undefined;
  }

  static extractClipData(data) {
    return data.items.map(({ id, snippet, statistics }) => {
      const {
        title, channelTitle, description, publishedAt,
      } = snippet;
      const { viewCount } = statistics;
      const descript = description.length > 150 ? `${description.slice(0, 150)}...` : description;
      const titleClip = title.length > 45 ? `${title.slice(0, 45)}...` : title;
      const channelName = channelTitle.length > 25 ? `${channelTitle.slice(0, 25)}...` : channelTitle;
      const image = snippet.thumbnails.medium;
      return {
        titleClip, channelName, descript, publishedAt, viewCount, image, id,
      };
    });
  }

  async getData() {
    const { urlApi, keyApi, request } = this.state;
    const nextPage = this.pageToken === undefined ? '' : `&pageToken=${this.pageToken}`;
    const url = `${urlApi}search?key=${keyApi}&type=video&part=snippet&maxResults=15&q=${request}${nextPage}`;
    const response = await fetch(url);
    const data = await response.json();

    const id = data.items.map(clip => clip.id.videoId).join(',');
    const { nextPageToken } = data;
    this.pageToken = nextPageToken;

    const statisticsUrl = `${urlApi}videos?key=${keyApi}&id=${id}&part=snippet,statistics`;
    const statisticsResponse = await fetch(statisticsUrl);
    const statisticsData = await statisticsResponse.json();

    return AppModel.extractClipData(statisticsData);
  }
}
