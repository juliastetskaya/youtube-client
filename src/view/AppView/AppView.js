import createElement from '../DomApi';

export default class AppView {
  static render(message) {
    const link = document.createElement('link');
    link.setAttribute('href', 'https://fonts.googleapis.com/css?family=PT+Sans:400,700');
    link.setAttribute('rel', 'stylesheet');

    const meta = document.createElement('meta');
    meta.setAttribute('name', 'viewport');
    meta.setAttribute('content', 'width=device-width,initial-scale=1');
    document.head.append(link, meta);

    const wrapper = createElement('div', 'page__wrapper');

    const inputSection = createElement('section', 'search');

    const input = createElement('input', 'search__box');
    input.setAttribute('placeholder', message);
    input.setAttribute('type', 'text');
    input.setAttribute('id', 'search-box');

    const label = createElement('label', 'search__label');
    label.setAttribute('for', 'search-box');

    label.append(input);
    inputSection.append(label);

    const clipSection = createElement('section', 'clip');

    const clipList = createElement('ul', 'clip__list');

    clipSection.append(clipList);

    wrapper.append(inputSection, clipSection);
    document.body.append(wrapper);
    document.body.classList.add('page');
  }

  static clearClips() {
    const clipList = document.querySelector('.clip__list');
    if (clipList !== null) {
      document.querySelector('.clip').removeChild(clipList);
      const newClipList = createElement('ul', 'clip__list');
      document.querySelector('.clip').append(newClipList);
    }
  }

  static renderClips(data) {
    data.forEach(({
      titleClip, descript, image, channelName, publishedAt, id, viewCount,
    }) => {
      const clipItem = createElement('li', 'clip__item');

      const clipWrapper = createElement('div', 'clip__wrapper');

      const clipImage = createElement('img', 'clip__image');
      clipImage.setAttribute('src', image.url);
      clipImage.setAttribute('alt', titleClip);
      clipImage.setAttribute('width', image.width);

      const clipLink = createElement('a', 'clip__link', titleClip);
      clipLink.setAttribute('href', `https://www.youtube.com/watch?v=${id}`);
      clipLink.setAttribute('target', '_blank');

      const clipTitle = createElement('h2', 'clip__title');
      clipTitle.append(clipLink);

      const clipInfo = createElement('div', 'clip__info');

      const clipChannelTitle = createElement('p', 'clip__channel-title', channelName);

      const clipDate = createElement('time', 'clip__date', publishedAt.slice(0, 10));
      clipDate.setAttribute('datetime', publishedAt);

      const clipViewCount = createElement('p', 'clip__view-count', viewCount);

      clipInfo.append(clipChannelTitle, clipDate, clipViewCount);

      const clipDescription = createElement('p', 'clip__description', descript);

      clipWrapper.append(clipImage, clipTitle, clipInfo, clipDescription);
      clipItem.append(clipWrapper);

      const clipList = document.querySelector('.clip__list');

      clipList.append(clipItem);
    });
  }

  static renderError() {
    const sectionError = createElement('section', 'error visually-hidden');
    const textError = createElement('p', 'error__text', 'No results found');

    sectionError.append(textError);

    document.querySelector('.page__wrapper').append(sectionError);
  }
}
