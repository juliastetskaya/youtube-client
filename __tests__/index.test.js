import AppView from '../src/view/AppView';
import AppModel from '../src/model/AppModel';
import createElement from '../src/view/DomApi';
import data from './__fixtures__/data';
import dataForRender from './__fixtures__/dataForRender';


describe('AppView', () => {
  it('Should be a function', () => {
    expect(AppView.render).toBeInstanceOf(Function);
  });

  it('Should be render correctly', () => {
    AppView.render('Search...');
    AppView.renderClips(dataForRender);

    expect(document.body.innerHTML).toMatchSnapshot();
  });
});

describe('AppModel', () => {
  it('ExtractClipData should be a function', () => {
    expect(AppModel.extractClipData).toBeInstanceOf(Function);
  });

  it('Result of work extractClipData', () => {
    const result = AppModel.extractClipData(data);
    expect(result[0].titleClip).toEqual('title 1');
    expect(result[1].channelName).toEqual('channel 1');
    expect(result[2].descript).toHaveLength(153);
    expect(result[1].publishedAt).toEqual('2015-12-11T03:30:00.000Z');
    expect(result[2].viewCount).toEqual('10453000');
    expect(result[0].image.url).toEqual('https://i.ytimg.com/vi/asnp3xm_vfY/mqdefault.jpg');
    expect(result[1].id).toEqual(2);
  });
});

describe('createElement', () => {
  it('test 1', () => {
    const expected = '<div></div>';
    expect(createElement('div').outerHTML).toEqual(expected);
  });

  it('test 2', () => {
    const expected = '<ul class="clip__list"></ul>';
    expect(createElement('ul', 'clip__list').outerHTML).toEqual(expected);
  });

  it('test 3', () => {
    const expected = '<p class="text">Some text</p>';
    expect(createElement('p', 'text', 'Some text').outerHTML).toEqual(expected);
  });

  it('test 4', () => {
    const expected = '<div class="block block-1">Some text</div>';
    expect(createElement('div', 'block block-1', 'Some text').outerHTML).toEqual(expected);
  });
});
