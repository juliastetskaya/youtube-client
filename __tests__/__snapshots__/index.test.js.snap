// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppView Should be render correctly 1`] = `"<div class=\\"page__wrapper\\"><section class=\\"search\\"><label class=\\"search__label\\" for=\\"search-box\\"><input class=\\"search__box\\" placeholder=\\"Search...\\" type=\\"text\\" id=\\"search-box\\"></label></section><section class=\\"clip\\"><ul class=\\"clip__list\\"><li class=\\"clip__item\\"><div class=\\"clip__wrapper\\"><img class=\\"clip__image\\" src=\\"https://i.ytimg.com/vi/McEFWLpZrnI/mqdefault.jpg\\" alt=\\"Exhaust worth Rs. 70,000 is now wasted?\\" width=\\"320\\"><h2 class=\\"clip__title\\"><a class=\\"clip__link\\" href=\\"https://www.youtube.com/watch?v=McEFWLpZrnI\\" target=\\"_blank\\">Exhaust worth Rs. 70,000 is now wasted?</a></h2><div class=\\"clip__info\\"><p class=\\"clip__channel-title\\">JS Films</p><time class=\\"clip__date\\" datetime=\\"2019-05-18T03:30:00.000Z\\">2019-05-18</time><p class=\\"clip__view-count\\">223373</p></div><p class=\\"clip__description\\">Get my Sunglasses from EyewearLabs: https://goo.gl/8oZwgH↵↵***Buy my MERCHANDISE***: https://bit.ly/,2DedHWW↵↵OFFICIAL WEBSITE: https://jaysnnation.com...</p></div></li><li class=\\"clip__item\\"><div class=\\"clip__wrapper\\"><img class=\\"clip__image\\" src=\\"https://i.ytimg.com/vi/nub1KpCQmWk/mqdefault.jpg\\" alt=\\"Уроки JavaScript Практика #7 Анимация в js\\" width=\\"320\\"><h2 class=\\"clip__title\\"><a class=\\"clip__link\\" href=\\"https://www.youtube.com/watch?v=nub1KpCQmWk\\" target=\\"_blank\\">Уроки JavaScript Практика #7 Анимация в js</a></h2><div class=\\"clip__info\\"><p class=\\"clip__channel-title\\">Web Developer Blog</p><time class=\\"clip__date\\" datetime=\\"2016-12-19T22:29:14.000Z\\">2016-12-19</time><p class=\\"clip__view-count\\">32468</p></div><p class=\\"clip__description\\">В этом уроке мы подробно разберем один из видов анимации в  языке программирования javascript.↵Ссылка на информацию: https://learn.javascript.ru/js-an...</p></div></li><li class=\\"clip__item\\"><div class=\\"clip__wrapper\\"><img class=\\"clip__image\\" src=\\"https://i.ytimg.com/vi/B7-yxJK3sfw/mqdefault.jpg\\" alt=\\"Зачем нужен Node js. Фреймворки Meteor, Expre...\\" width=\\"320\\"><h2 class=\\"clip__title\\"><a class=\\"clip__link\\" href=\\"https://www.youtube.com/watch?v=B7-yxJK3sfw\\" target=\\"_blank\\">Зачем нужен Node js. Фреймворки Meteor, Expre...</a></h2><div class=\\"clip__info\\"><p class=\\"clip__channel-title\\">Web Developer Blog</p><time class=\\"clip__date\\" datetime=\\"2018-08-03T13:57:13.000Z\\">2018-08-03</time><p class=\\"clip__view-count\\">16302</p></div><p class=\\"clip__description\\">В этом видео мы разберем частый вопрос от начинающих зачем нужен Node js. Ведь новички до конца не понимают как Node js используется в проектах, за сч...</p></div></li></ul></section></div>"`;
